expertsystem/data.py

```diff
@@ -84,29 +99,121 @@ def __hash__(self) -> int:
         return hash(repr(self))
 
 
-class Particle(NamedTuple):
-    """Immutable data container for particle info."""
+@dataclass(frozen=True)
+class HasQuantumNumbers:  # pylint: disable=too-many-instance-attributes
+    """Set of quantum numbers **excluding spin**.
 
-    name: str
-    pid: int
-    charge: int
-    spin: float
-    mass: float
+    This is to make spin projection required in `.QuantumState` (`.HasSpin`)
+    and unavailable in `.Particle` (`.HasFloatSpin`).
+    """
+
+    charge: int = 0
+    isospin: Optional[Spin] = None
     strangeness: int = 0
     charmness: int = 0
     bottomness: int = 0
     topness: int = 0
     baryon_number: int = 0
-    electron_number: int = 0
-    muon_number: int = 0
-    tau_number: int = 0
-    width: Optional[float] = None
-    isospin: Optional[Spin] = None
+    electron_lepton_number: int = 0
+    muon_lepton_number: int = 0
+    tau_lepton_number: int = 0
     parity: Optional[Parity] = None
     c_parity: Optional[Parity] = None
     g_parity: Optional[Parity] = None
 
 
+@dataclass(frozen=True)
+class HasSpin:
+    """Required to disallow default arguments for `.QuantumState`."""
+
+    spin: Spin
+
+
+@dataclass(frozen=True)
+class HasFloatSpin:
+    """Required to disallow default arguments for `.ParticleQuantumState`."""
+
+    spin: float
+
+
+@dataclass(frozen=True)
+class QuantumState(HasQuantumNumbers, HasSpin):
+    """Contains all quantum numbers unambiguously defining a quantum state."""
+
+
+@dataclass(frozen=True)
+class ParticleQuantumState(HasQuantumNumbers, HasFloatSpin):
+    """Similar to `.QuantumState` but only carrying spin magnitude."""
+
+
+class HasComplexEnergy:
+    """Required to disallow default arguments for `.ComplexEnergyState`."""
+
+    def __init__(self, energy: complex):
+        self.__energy = complex(energy)
+
+    @property
+    def complex_energy(self) -> complex:
+        return self.__energy
+
+    @property
+    def mass(self) -> float:
+        return self.__energy.real
+
+    @property
+    def width(self) -> float:
+        return self.__energy.imag
+
+
+class ComplexEnergyState(HasComplexEnergy):
+    """Pole in the complex energy plane, with quantum numbers."""
+
+    def __init__(self, energy: complex, state: QuantumState):
+        super().__init__(energy)
+        self.state: QuantumState = state
```

Comment:
To filter out double solutions later on, we will also need the `__eq__` operation here.

Reply:
Better to implement that once there's a need for it, then it's immediately tested as well.